File: lib/src/application/use_cases/log.dart
import 'package:functionally/extensions/reader_io.dart';
import 'package:functionally/reader_io.dart' as RIO;

import '../commands/log.dart' as logCommand;

enum LogType { debug, info, error }

typedef Log = void Function(LogType, String);

RIO.ReaderIO<logCommand.LogCommand, void> prepareLog(
  LogType type,
  String message,
) =>
    RIO.ask<logCommand.LogCommand>().flatMapIO(
          (log) => log(
            switch (type) {
              LogType.debug => logCommand.LogType.debug,
              LogType.error => logCommand.LogType.error,
              LogType.info => logCommand.LogType.info
            },
            message,
          ),
        );

File: src/app/commands/log.ts
import * as IO from 'fp-ts/IO'
import * as RIO from 'fp-ts/ReaderIO'

export type LogType =
	| 'debug'
	| 'info'
	| 'warning'
	| 'error'

export type Log = (
	type: LogType,
	message: string,
) => IO.IO<void>

export type R_Log<ENV> = (
	type: LogType,
	message: string,
) => RIO.ReaderIO<ENV, void>
